File: templates/problems/13_dp_1d/01_climbing_stairs/solution.go
package climbing_stairs

// solveClimbStairs returns the number of ways to climb n stairs. O(n) time, O(1) space.
func solveClimbStairs(n int) int {
	if n <= 2 {
		return n
	}
	prev, curr := 1, 2
	for i := 3; i <= n; i++ {
		prev, curr = curr, prev+curr
	}
	return curr
}

File: templates/problems/13_dp_1d/01_climbing_stairs/typescript/solution.ts
/**
 * Return the number of ways to climb n stairs.
 * Time: O(n), Space: O(1)
 */
export function solveClimbStairs(n: number): number {
    if (n <= 2) return n;
    let prev = 1;
    let curr = 2;
    for (let i = 3; i <= n; i++) {
        [prev, curr] = [curr, prev + curr];
    }
    return curr;
}
